### PR TITLE
Feature/storage market logic

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
### Description of the Change
Add state handlers for storage market.
- check signatures
- check proposal
- check publish deal
- clean up on final state
- enable tests
Reference: https://github.com/filecoin-project/go-fil-markets/tree/master/storagemarket/impl/providerstates

### CODE REVIEW

File: core/markets/storage/client/impl/storage_market_client_impl.cpp [hunk 1]
```diff
@@ -481,16 +573,21 @@ namespace fc::markets::storage::client {
...
+    FSM_SEND(deal, ClientEvent::ClientEventDealPublished);
   }
 
   void StorageMarketClientImpl::onClientEventDealPublished(
       std::shared_ptr<ClientDeal> deal,
       ClientEvent event,
       StorageDealStatus from,
       StorageDealStatus to) {
-    // verify deal activated - on deal sector commit
+    // TODO (a.chernyshov) verify deal activated - on deal sector commit
```
Comment (igor-egorov) [inline core/markets/storage/client/impl/storage_market_client_impl.cpp:590]: why not implement this now?

File: core/markets/storage/client/impl/storage_market_client_impl.hpp [hunk 2]
```diff
@@ -279,7 +304,8 @@ namespace fc::markets::storage::client {
     PIECE_DATA_NOT_SET_MANUAL_TRANSFER,
     PIECE_SIZE_GREATER_SECTOR_SIZE,
     ADD_FUNDS_CALL_ERROR,
-    LOCAL_DEAL_NOT_FOUND
+    LOCAL_DEAL_NOT_FOUND,
+    STREAM_LOOKUP_ERROR
```
Comment (igor-egorov) [inline core/markets/storage/client/impl/storage_market_client_impl.hpp:308]: very minor, mostly stylistic. Better to have `,` after the last element.

File: core/markets/storage/common.hpp [hunk 3]
```diff
@@ -6,17 +6,8 @@
 #ifndef CPP_FILECOIN_MARKETS_STORAGE_COMMON_HPP
 #define CPP_FILECOIN_MARKETS_STORAGE_COMMON_HPP
 
-#include "primitives/types.hpp"
+#define FSM_SEND(deal, event) OUTCOME_EXCEPT(fsm_->send(deal, event))
```
Comment (igor-egorov) [inline core/markets/storage/common.hpp:9]: Is it safe to use here `OUTCOME_EXCEPT` which may raise an exception?

File: core/markets/storage/provider/impl/provider_impl.cpp [hunk 4]
```diff
@@ -346,26 +330,43 @@ namespace fc::markets::storage::provider {
...
+                                   .signature = signature};
+    OUTCOME_TRY(stream, getStream(deal->proposal_cid));
+    stream->write(signed_response,
+                  [self{shared_from_this()}, stream, deal](
+                      outcome::result<size_t> maybe_res) {
+                    if (maybe_res.has_error()) {
+                      // assume client disconnected
+                      self->logger_->error("Write deal response error. "
+                                           + maybe_res.error().message());
+                      return;
```
Comment (igor-egorov) [inline core/markets/storage/provider/impl/provider_impl.cpp:362]: should we update the deal state before returning?

File: core/markets/storage/provider/storage_provider_error.hpp [hunk 5]
```diff
@@ -11,7 +11,8 @@
 namespace fc::markets::storage::provider {
 
   enum class StorageProviderError {
-    PROVIDER_START_ERROR = 1
+    PROVIDER_START_ERROR = 1,
+    STREAM_LOOKUP_ERROR
```
Comment (igor-egorov) [inline core/markets/storage/provider/storage_provider_error.hpp:15]:
```suggestion
    STREAM_LOOKUP_ERROR,
```

File: test/core/markets/storage/storage_market_ask.cpp [hunk 6]
```diff
@@ -18,7 +18,7 @@ namespace fc::markets::storage::test {
    * @when client send get ask
    * @then ask returned in answer
    */
-  TEST_F(StorageMarketTest, DISABLED_Ask) {
+  TEST_F(StorageMarketTest, Ask) {
```
Comment (igor-egorov) [inline test/core/markets/storage/storage_market_ask.cpp:21]: 👍

File: core/markets/storage/client/impl/storage_market_client_impl.cpp (hunk 1 above)
Comment (Alexey-N-Chernyshov) [reply core/markets/storage/client/impl/storage_market_client_impl.cpp:590]: There is no sector builder and api method for now.

File: core/markets/storage/common.hpp (hunk 3 above)
Comment (Alexey-N-Chernyshov) [reply core/markets/storage/common.hpp:9]: If FSM failed, we can only report problem and stop execution.

File: core/markets/storage/provider/impl/provider_impl.cpp (hunk 4 above)
Comment (Alexey-N-Chernyshov) [reply core/markets/storage/provider/impl/provider_impl.cpp:362]: It is updated in CALLBACK_ACTION.

**Pull request merged**